new-tutorial/L4_message_passing.py

```diff
@@ -0,0 +1,386 @@
+    graph, train_nids, sampler,
+    batch_size=1024,
+    shuffle=True,
+    drop_last=False,
+    num_workers=0
+)
+
+input_nodes, output_nodes, bipartites = next(iter(train_dataloader))
+
+
+######################################################################
+# DGL Bipartite Graph Introduction
+# --------------------------------
+# 
+# In the previous tutorials, you have seen the concept *bipartite graph*.
+# This section introduces how you can manipulate (directional) bipartite
+# graphs.
+# 
+# You can access the input node features and output node features via
+# ``srcdata`` and ``dstdata`` attributes:
+# 
+
+bipartite = bipartites[0]
+print(bipartite.srcdata)
+print(bipartite.dstdata)
+
+
+######################################################################
+# It also has ``num_src_nodes`` and ``num_dst_nodes`` functions to query
+# how many input nodes and output nodes exist in the bipartite graph:
+# 
+
+print(bipartite.num_src_nodes(), bipartite.num_dst_nodes())
+
+
+######################################################################
+# You can assign features to ``srcdata`` and ``dstdata`` just as what you
+# will do with ``ndata`` on the graphs you have seen earlier:
+# 
+
+bipartite.srcdata['x'] = torch.zeros(bipartite.num_src_nodes(), bipartite.num_dst_nodes())
+dst_feat = bipartite.dstdata['feat']
+
+
+######################################################################
+# Also, since the bipartite graphs are constructed by DGL, you can
+# retrieve the input node IDs (i.e. those that are required to compute the
+# output) and output node IDs (i.e. those whose representations the
+# current GNN layer should compute) as follows.
+# 
+
+bipartite.srcdata[dgl.NID], bipartite.dstdata[dgl.NID]
+
+
+######################################################################
+# Writing GNN Modules for Bipartite Graphs for Stochastic Training
+# ----------------------------------------------------------------
+# 
+
+
+######################################################################
+# Recall that the bipartite graphs yielded by the ``NodeDataLoader`` and
+# ``EdgeDataLoader`` have the property that the first few input nodes are
+# always identical to the output nodes:
+# 
+# |image1|
+# 
+# .. |image1| image:: https://data.dgl.ai/tutorial/img/bipartite.gif
+# 
+
+print(torch.equal(bipartite.srcdata[dgl.NID][:bipartite.num_dst_nodes()], bipartite.dstdata[dgl.NID]))
+
+
+######################################################################
+# Suppose you have obtained the input node representations
+# :math:`h_u^{(l-1)}`:
+# 
+
+bipartite.srcdata['h'] = torch.randn(bipartite.num_src_nodes(), 10)
+
+
+######################################################################
+# This means that the input nodes is a union of the output nodes and their
+# neighbors, enabling you to conveniently get the term :math:`h_v^{(l-1)}`
+# via:
+# 
+
+h_v = bipartite.srcdata['h'][:bipartite.num_dst_nodes()]
```

Comment:
Shall we mention this slicing behavior again? This is about the change of model not GNN module right?

Reply:
No this is no longer necessary.